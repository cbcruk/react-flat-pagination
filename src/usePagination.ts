import { useCallback, useState } from 'react'
import { chunk, range } from './utils'
import { Options } from './types'

export function usePagination({
  totalNumber,
  pageSize,
  pageRange = 10,
  pageNumber = 1,
}: Options) {
  const [current, setCurrent] = useState(pageNumber)
  const list = range(Math.ceil(totalNumber / pageSize))
  const chunkCount = totalNumber / pageRange
  const chunked = chunk({ list, count: chunkCount })
  const position = Math.ceil(current / pageRange) - 1
  const pagination = chunked[position]
  const handlePrev = useCallback(() => setCurrent((p) => p - 1), [])
  const handleNext = useCallback(() => setCurrent((p) => p + 1), [])
  const goTo = useCallback((index: number) => setCurrent(index), [])
  const goToFirst = useCallback(() => setCurrent(1), [])
  const goToLast = useCallback(() => setCurrent(list.length), [])

  return {
    pagination,
    current,
    setCurrent,
    handlePrev,
    handleNext,
    goTo,
    goToLast,
    goToFirst,
  }
}
