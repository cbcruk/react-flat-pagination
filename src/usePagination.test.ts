import { expect, test } from 'vitest'
import { renderHook } from '@testing-library/react-hooks'
import { usePagination } from './usePagination'

test('usePagination', () => {
  const { result } = renderHook(() =>
    usePagination({
      totalNumber: 100,
      pageSize: 30,
    })
  )

  expect(result.current).toMatchInlineSnapshot(`
    {
      "current": 1,
      "goTo": [Function],
      "goToFirst": [Function],
      "goToLast": [Function],
      "handleNext": [Function],
      "handlePrev": [Function],
      "pagination": [
        1,
        2,
        3,
        4,
      ],
      "setCurrent": [Function],
    }
  `)

  result.current.setCurrent(2)
  expect(result.current.current).toEqual(2)

  result.current.goTo(3)
  expect(result.current.current).toEqual(3)

  result.current.goToFirst()
  expect(result.current.current).toEqual(1)

  result.current.handleNext()
  expect(result.current.current).toEqual(2)

  result.current.handlePrev()
  expect(result.current.current).toEqual(1)

  result.current.goToLast()
  expect(result.current.current).toEqual(4)
})
