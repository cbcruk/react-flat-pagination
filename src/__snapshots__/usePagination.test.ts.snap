// Vitest Snapshot v1

exports[`usePagination 2`] = `
{
  "current": 2,
  "goTo": [Function],
  "goToFirst": [Function],
  "goToLast": [Function],
  "handleNext": [Function],
  "handlePrev": [Function],
  "pagination": [
    1,
    2,
    3,
  ],
  "setCurrent": [Function],
}
`;
